R/facet-.R

```diff
@@ -88,7 +88,95 @@ Facet <- ggproto("Facet", NULL,
     cli::cli_abort("Not implemented.")
   },
   map_data = function(data, layout, params) {
-    cli::cli_abort("Not implemented.")
+
+    if (empty(data)) {
+      return(vec_cbind(data %|W|% NULL, PANEL = integer(0)))
+    }
+
+    vars <- params$facet %||% c(params$rows, params$cols)
+
+    if (length(vars) == 0) {
+      data$PANEL <- layout$PANEL
+      return(data)
+    }
+
+    grid_layout <- all(c("rows", "cols") %in% names(params))
+    layer_layout <- attr(data, "layout")
+    if (identical(layer_layout, "fixed")) {
+      n <- vec_size(data)
+      data <- vec_rep(data, vec_size(layout))
+      data$PANEL <- vec_rep_each(layout$PANEL, n)
+      return(data)
+    }
+
+    # Compute faceting values
+    facet_vals <- eval_facets(vars, data, params$.possible_columns)
+
+    include_margins <- !isFALSE(params$margin %||% FALSE) &&
+      nrow(facet_vals) == nrow(data) && grid_layout
+    if (include_margins) {
+      # Margins are computed on evaluated faceting values (#1864).
```

Comment:
I merged the wrap/grid layout mapping because it reduces duplication and the only tangible difference was dealing with margins.